sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SpecifyIFormatProviderOrCultureInfo.cs

```diff
@@ -92,18 +92,99 @@ protected override void Initialize(SonarAnalysisContext context)
+        {
+            var parametersWithoutFormatCulture = possibleOverload.GetParameters().Where(p => !p.Type.IsAny(formatAndCultureType));
+            // no FormatOrCulture argument found
+            if (parametersWithoutFormatCulture.Count() == possibleOverload.GetParameters().Count())
+            {
+                return false;
+            }
+
+            // once we filter out the FormatOrCulture argument, we have two possibilities:
+            // - possibleOverload has a 'params' argument which matches the invocationParameters
+            // - the number of parameters is the same and of same type
+
+            var invocationParametersNumber = invocationParameters.Count();
+            var parametersWithoutFormatCultureNumber = parametersWithoutFormatCulture.Count();
+
+            if (parametersWithoutFormatCultureNumber <= invocationParametersNumber &&
+                parametersWithoutFormatCultureNumber > 0 &&
+                parametersWithoutFormatCulture.Last() is IParameterSymbol lastParameter &&
+                lastParameter.IsParams)
+            {
+                return VerifyCompatibility(invocationParameters.ToList(), parametersWithoutFormatCulture.ToList(), lastParameter);
+            }
+            else if (invocationParametersNumber == parametersWithoutFormatCultureNumber)
+            {
+                return VerifyCompatibility(invocationParameters.ToList(), parametersWithoutFormatCulture.ToList());
+            }
+            return false;
+        }
+
+        /**
+         * Verifies the compatibility between the invocation parameters and the parameters of a possible overload that
+         * has the last parameter of 'params' type (variable length parameter).
+         */ 
+        private static bool VerifyCompatibility(IList<IParameterSymbol> invocationParameters,
+            IList<IParameterSymbol> overloadCandidateParameters,
+            IParameterSymbol paramsParameter)
+        {
+            var i = 0;
+            // check parameters before the last parameter
+            for (; i < overloadCandidateParameters.Count - 1; i++)
+            {
+                if (!invocationParameters[i].Type.DerivesOrImplements(overloadCandidateParameters[i].Type))
+                {
+                    return false;
+                }
+            }
+            // make sure the rest of the invocation parameters match with the 'params' type
+            var paramsType = GetParamsType(paramsParameter.Type);
+            for (; i < invocationParameters.Count; i++)
+            {
+                if (!invocationParameters[i].Type.DerivesOrImplements(paramsType))
+                {
+                    return false;
+                }
+            }
+            return true;
+
+            ITypeSymbol GetParamsType(ITypeSymbol typeSymbol) => (typeSymbol as IArrayTypeSymbol)?.ElementType;

```

Comment:
This line is coded defensively to handle `typeSymbol` not being an array type. If we think this could happen in practice (e.g. we're passed an invalid syntax tree that somehow gives a valid symbol that is a `params` symbol but not an array type) then it would be better to do the check right at the start of the method and exit early if we can't work out the type.
If we don't think it could happen in practice then the defensive code isn't needed.